### PR TITLE
Save interfaces implementations and nested types

### PR DESCRIPTION
**Pull request by buyaa-n**
Contributes to https://github.com/dotnet/runtime/issues/62956

Saves interfaces implemented and nested types

### CODE REVIEW

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveWithVariousMembersTests.cs [hunk 1]
```diff
@@ -122,22 +123,52 @@ public void CreateMembersThatUsesTypeLoadedFromCoreAssemblyTest()
...
+        }
     }
 
     // Test Types
     public interface INoMethod
     {
     }
 
-    public interface IMultipleMethod
+    public interface IMultipleMethod : IEnumerable
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveWithVariousMembersTests.cs:163]:
```suggestion
    public interface IMultipleMethod
```

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveWithVariousMembersTests.cs [hunk 2]
```diff
@@ -122,22 +123,52 @@ public void CreateMembersThatUsesTypeLoadedFromCoreAssemblyTest()
...
     {
         string Func(int a, string b);
         IOneMethod MoreFunc();
         StructWithFields DoIExist(int a, string b, bool c);
         void BuildAPerpetualMotionMachine();
     }
 
-    internal interface IAccess
+    internal interface IAccess : IOneMethod
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveWithVariousMembersTests.cs:171]:
```suggestion
    internal interface IAccess
```

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs [hunk 3]
```diff
@@ -132,6 +132,25 @@ internal void AppendMetadata()
                 Debug.Assert(typeBuilder._handle.Equals(typeDefinitionHandle));
                 WriteCustomAttributes(typeBuilder._customAttributes, typeDefinitionHandle);
 
+                if (typeBuilder.Attributes.HasFlag(TypeAttributes.ExplicitLayout))
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:135]: Slightly better perf characteristics. Also, it is what we use on in other places in this file.

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs [hunk 4]
```diff
@@ -45,12 +48,27 @@ internal sealed class TypeBuilderImpl : TypeBuilder
...
             }
+
+            if (interfaces != null)
+            {
+                _interfaces = new List<Type>();
+                for (int i = 0; i < interfaces.Length; i++)
+                {
+                    // cannot contain null in the interface list
+                    ArgumentNullException.ThrowIfNull(interfaces[i], nameof(interfaces));
+                    _interfaces.Add(interfaces[i]);
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs:59]:
```suggestion
                    Type interface = interfaces[i];
                    ArgumentNullException.ThrowIfNull(interface, nameof(interfaces));
                    _interfaces.Add(interface);
```
Nit: Makes sure that the null won't sneak in if the caller modifies the array on a second thread.

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs [hunk 5]
```diff
@@ -17,18 +17,20 @@ internal sealed class TypeBuilderImpl : TypeBuilder
...
         private Type? _typeParent;
+        private readonly TypeBuilderImpl? _declaringType;
         private TypeAttributes _attributes;
         private PackingSize _packingSize;
         private int _typeSize;
 
         internal readonly TypeDefinitionHandle _handle;
         internal readonly List<MethodBuilderImpl> _methodDefinitions = new();
         internal readonly List<FieldBuilderImpl> _fieldDefinitions = new();
+        internal List<Type>? _interfaces;
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs:28]: You can also implement `GetInterfaces` method while you are on it.

**Review by jkotas — APPROVED** [review]

**Pull request merged**